File: doing_things_with_data/doing_things_with_Py/scripts_py/analyzing_data/linear_regressions/simple_linear_regression_manual_calculation.py
####simple linear regression manual calculations####

###data
x = [7,10,15,23,6,23,24,17,16,2]
y = [10,12,22,27,14,28,32,27,20,10]

###descriptive statistic calculations
##mean values for predictor and outcome variables 
x_mean = sum(x)/len(x)
y_mean = sum(y)/len(y)

##deviation scores
#x deviation scores
x_dev = [i - x_mean for i in x]
#y deviation scores
y_dev = [i - y_mean for i in y]
#x deviation scores squared
x_dev2 = [i**2 for i in x_dev]
#y deviation scores squared
y_dev2 = [i**2 for i in y_dev]
#product of x deviation and y deviation scores
xy_dev = [xvalue * y_dev[num] for num, xvalue in enumerate(x_dev)]

###estimating the coefficients
#slope
b1 = sum(xy_dev)/sum(x_dev2)
#intercept
b0 = y_mean - (b1 * x_mean)


###evaluating the model
##predicted values and error
#predicted values based on the estimated coefficients
y_pred = [b0 + (b1 * i) for i in x] 
#residuals: difference between the observed outcome scores and the predicted values
residuals = [yvalue - y_pred[num] for num, yvalue in enumerate(y)]
#residuals sqaured 
residuals2 = [i ** 2 for i in residuals]
#model error
model_error = [i - y_mean for i in y_pred]
#model error sqaured
model_error2 = [i ** 2 for i in model_error] 

##overall model fit
#residual sum of squares
rss = sum(residuals2)
#model sum of squares
mss = sum(model_error2)
#total sum of squares
tss = rss + mss
#residual standard error
rse = (rss/(len(y)-2)) ** (1/2.0)
#r-squared value
r_sq = (tss - rss)/tss
#standard error of intercept and slope 
b0_se = rse * (((1/len(x)) + (x_mean**2/sum(x_dev2))) ** (1/2.0))
b1_se = rse/((sum(x_dev2) ** (1/2.0)))
#f-statistic
f = (mss/1)/(rss/(len(x)-2))
#with simiple linear regressions the f-statistic can be found by squaring the t-statistic for the slope

####assessing predictive value
b0_t = (b0 - 0)/b0_se
b1_t = (b1 - 0)/b1_se


#predicted outcome based on new predictor value (e.g. 12)
new_y1 = b0 + (b1 * 12)

File: doing_things_with_data/doing_things_with_R/scripts_r/analyzing_data/linear_regressions/simple_linear_regression_manual_calculation.R
####simple linear regression (manual calculations)####

###data 
set.seed(1) #set seed for reproducibility 
x <- sample(x = 1:25, size = 10, replace = T) #predictor values 
y <- x + sample(x = 1:10, size = 10, replace = T) #outcome values 
df <- data.frame(x, y)
df

###descriptive statistic calculations 
##mean values for predictor and outcome variables
x_mean <- mean(df$x)
y_mean <- mean(df$y)

##deviance scores 
df$x_dev <- df$x - x_mean #x deviance scores 
df$y_dev <- df$y - y_mean #y deviance scores 
df$x_dev2 <- df$x_dev^2 #x deviance scores sqaured 
df$y_dev2 <- df$y_dev^2 #y deviance scores sqaured
df$xy_dev <- df$x_dev * df$y_dev #product of x deviance and y deviance scores 

###estimating the coefficients 
b1 <- sum(df$xy_dev)/sum(df$x_dev2)
b0 <- y_mean - (b1 * x_mean)

###evaluating the model 
##predicted values and error  
df$y_pred <- b0 + (b1 * df$x) #predicted values based on the estimated coefficients 
df$residuals <- df$y - df$y_pred #residuals: difference between the observed outcome scores and the predicted values 
df$residuals2 <- df$residuals^2 #residuals sqaured 
df$model_error <- df$y_pred - y_mean #model error 
df$model_error2 <- df$model_error^2 #model error sqaured 

##overall model fit 
rss <- sum(df$residuals2) #residual sum of sqaures 
mss <- sum(df$model_error2) #model sum of sqaures 
tss <- rss + mss #total sum of sqaures; or sum(df$y_dev2), the summ of sqaured outcome deviation scores 
rse <- sqrt(rss/(length(df$y)-2)) #residual standard error 
r_sq <- (tss - rss)/tss #r-squared value 

b0_se <- rse * sqrt((1/length(df$x)) + (x_mean^2/sum(df$x_dev2))) #standard error of the intercept
b1_se <- rse/sqrt(sum(df$x_dev2)) #standard error of the slope 

###assessing predictive value
b0_t <- (b0 - 0)/b0_se
b0_t_pval <- 2 * pt(-abs(b0_t), df = length(df$x)-2)
b1_t <- (b1 - 0)/b1_se
b1_t_pval <- 2 * pt(-abs(b1_t), df = length(df$x)-2)

#predicted outcome based on new predictor value (e.g. 12)
new_y1 <- b0 + (b1 * 12)


